File: 0003-longest-substring-without-repeating-characters/0003-longest-substring-without-repeating-characters.swift
class Solution {
    func lengthOfLongestSubstring(_ s: String) -> Int {
        // Map from character to its last seen index
        var lastSeen: [Character: Int] = [:]
        
        let chars = Array(s)
        var left = 0          // left bound of current window
        var maxLen = 0
        
        for right in 0..<chars.count {
            let c = chars[right]
            
            if let prevIndex = lastSeen[c], prevIndex >= left {
                // Move left just past the previous occurrence of this character
                left = prevIndex + 1
            }
            
            // Update last seen index for this character
            lastSeen[c] = right
            
            // Window length = right - left + 1
            maxLen = max(maxLen, right - left + 1)
        }
        
        return maxLen
    }
}

File: 0003-longest-substring-without-repeating-characters/0003-longest-substring-without-repeating-characters.cpp
class Solution {
public:
    int lengthOfLongestSubstring(string s) {
        vector<int> lastIndex(256, -1); // store last index of each char
        int maxLen = 0;
        int left = 0;                   // start of current window

        for (int right = 0; right < (int)s.size(); ++right) {
            char c = s[right];

            // if char seen and is inside current window, move left
            if (lastIndex[c] >= left) {
                left = lastIndex[c] + 1;
            }

            lastIndex[c] = right;               // update last seen index
            maxLen = max(maxLen, right - left + 1);
        }

        return maxLen;
    }
};
